### ISSUE TITLE
Segmentation fault running inference on gemma3

### ISSUE THREAD

**Issue by MuradAlmakhzangi** [google-ai-edge/mediapipe#6074]
### Have I written custom code (as opposed to using a stock example script provided in MediaPipe)

No

### OS Platform and Distribution

Android 15

### Task name (e.g. Image classification, Gesture recognition etc.)

LLM Inference Engine CPU main

### Programming Language and version (e.g. C++, Python, Java)

C++

### Describe the actual behavior

Segmentation Fault when running the cpu inference main task

### Describe the expected behaviour

Completion for the default prompt

### Standalone code/steps you may have used to try to get what you need

Building the task using  `bazel build --cxxopt=-DABSL_FLAGS_STRIP_NAMES=0  -c opt --config=android_arm64  mediapipe/tasks/cc/genai/inference/c:llm_inference_engine_cpu_main`

### Other info / Complete Logs

```shell
After pushing it to the device (Pixel 9) along with gemma3's .task file, running as such

tokay:/ $ ./tmp/user_0/llm/llm_inference_engine_cpu_main --model_path /tmp/user_0/llm/gemma3-1B-it-int4.task                                                                                                      
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1756820727.512966   19181 llm_inference_engine_cpu_main.cc:119] Prompt: Write an email
INFO: Initialized TensorFlow Lite runtime.
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
VERBOSE: XNNPack weight cache loaded from '/tmp/user_0/llm/gemma3-1B-it-int4.task.xnnpack_cache'.
VERBOSE: Replacing 1105 out of 1330 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 316 partitions for subgraph 0.
VERBOSE: Replacing 1153 out of 1373 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 307 partitions for subgraph 1.
VERBOSE: Replacing 1153 out of 1373 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 307 partitions for subgraph 2.
VERBOSE: Replacing 1153 out of 1373 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 307 partitions for subgraph 3.
VERBOSE: Replacing 1153 out of 1373 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 307 partitions for subgraph 4.

... 600+ lines of this

VERBOSE: Replacing 8 out of 8 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 1 partitions for subgraph 768.
VERBOSE: Replacing 8 out of 8 node(s) with delegate (TfLiteXNNPackDelegate) node, yielding 1 partitions for subgraph 769.
normalizer.cc(52) LOG(INFO) precompiled_charsmap is empty. use identity normalization.
I0000 00:00:1756820727.743758   19181 llm_inference_engine_cpu_main.cc:149] AddQueryChunk
I0000 00:00:1756820727.743802   19181 llm_inference_engine_cpu_main.cc:164] PredictAsync
I0000 00:00:1756820727.743897   19181 llm_inference_engine_cpu_main.cc:186] DeleteSession
Segmentation fault
```

**Comment by MuradAlmakhzangi** [comment]
Looking at the `start_llm_function` the signature runner is written as "prefill" but checking the possible ones for gemma3's .task it looks like it should be one of "prefill_[32,128,512,1024]"

Similar issues for input and decode tensors being retrieved as "args_0" and "args_1" possibly

**Comment by HarishPermal** [comment]
Hi @MuradAlmakhzangi,

Sorry for the delayed response, im able to reproduce, this issue from my end, i'll update you once i found any insights.

**Comment by HarishPermal** [comment]
Hi @MuradAlmakhzangi,

After looking at some crash logs from `adb logcat -b crash -d` i ran into the Segmentation fault, because of low memory on my device.

Below you can see

<img width="581" height="89" alt="Image" src="https://github.com/user-attachments/assets/4720b98e-2993-4f45-9b5f-f18e3fccc4c8" />

I would encourage you to share the crash logs by following the below steps

```bash
abd logcat -b crash -c

adb shell "./tmp/user_0/llm/llm_inference_engine_cpu_main --model_path /tmp/user_0/llm/gemma3-1B-it-int4.task"

adb logcat -b crash -d
```

Thank You

**Comment by MuradAlmakhzangi** [comment]
The output of the crash shows this
```
muralm01@CGJG3G7WWG ~ % adb logcat -b crash -d
--------- beginning of crash
09-25 08:51:44.400 28054 28059 F libc    : Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x8 in tid 28059 (llm_inference_e), pid 28054 (llm_inference_e)
09-25 08:51:44.537 28065 28065 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
09-25 08:51:44.537 28065 28065 F DEBUG   : Build fingerprint: 'google/tokay/tokay:15/BP1A.250505.005.A1/13277748:user/release-keys'
09-25 08:51:44.537 28065 28065 F DEBUG   : Revision: 'MP1.0'
09-25 08:51:44.537 28065 28065 F DEBUG   : ABI: 'arm64'
09-25 08:51:44.537 28065 28065 F DEBUG   : Timestamp: 2025-09-25 08:51:44.463172833+0100
09-25 08:51:44.537 28065 28065 F DEBUG   : Process uptime: 1s
09-25 08:51:44.537 28065 28065 F DEBUG   : Cmdline: ./tmp/user_0/llm/llm_inference_engine_cpu_main --model_path /tmp/user_0/llm/gemma3-1B-it-int4.task
09-25 08:51:44.537 28065 28065 F DEBUG   : pid: 28054, tid: 28059, name: llm_inference_e  >>> ./tmp/user_0/llm/llm_inference_engine_cpu_main <<<
09-25 08:51:44.537 28065 28065 F DEBUG   : uid: 2000
09-25 08:51:44.537 28065 28065 F DEBUG   : tagged_addr_ctrl: 0000000000000001 (PR_TAGGED_ADDR_ENABLE)
09-25 08:51:44.537 28065 28065 F DEBUG   : pac_enabled_keys: 000000000000000f (PR_PAC_APIAKEY, PR_PAC_APIBKEY, PR_PAC_APDAKEY, PR_PAC_APDBKEY)
09-25 08:51:44.537 28065 28065 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0000000000000008
09-25 08:51:44.537 28065 28065 F DEBUG   : Cause: null pointer dereference
09-25 08:51:44.537 28065 28065 F DEBUG   :     x0  0000000000000000  x1  0000005cc09bca2b  x2  006c6c6966657270  x3  617261705f6d6c6c
09-25 08:51:44.537 28065 28065 F DEBUG   :     x4  0000000000000038  x5  0000000000000000  x6  6071606f5e6c6b6b  x7  7f7f7f7f7f7f7f7f
09-25 08:51:44.537 28065 28065 F DEBUG   :     x8  000000000000000b  x9  000000003231355f  x10 000000000000000b  x11 000000000000000d
09-25 08:51:44.537 28065 28065 F DEBUG   :     x12 0000000000000000  x13 0000000000001000  x14 0000000000100000  x15 28f5c28f5c28f5c3
09-25 08:51:44.537 28065 28065 F DEBUG   :     x16 0000005cc0ef9280  x17 000000743505cec0  x18 0000007140e4c000  x19 b400007295f1c8d0
09-25 08:51:44.537 28065 28065 F DEBUG   :     x20 0000000000000000  x21 0000007142f06650  x22 0000000000006d96  x23 0000000000006d96
09-25 08:51:44.537 28065 28065 F DEBUG   :     x24 0000007142f06650  x25 0000007142f06650  x26 0000007142f06990  x27 b400007295f1c8d0
09-25 08:51:44.537 28065 28065 F DEBUG   :     x28 0000000000104000  x29 0000007142f06600
09-25 08:51:44.538 28065 28065 F DEBUG   :     lr  0000005cc0ac651c  sp  0000007142f06570  pc  0000005cc0ac9bc0  pst 0000000060001000
09-25 08:51:44.538 28065 28065 F DEBUG   : 4 total frames
09-25 08:51:44.538 28065 28065 F DEBUG   : backtrace:
09-25 08:51:44.538 28065 28065 F DEBUG   :   NOTE: Function names and BuildId information is missing for some frames due
09-25 08:51:44.538 28065 28065 F DEBUG   :   NOTE: to unreadable libraries. For unwinds of apps, only shared libraries
09-25 08:51:44.538 28065 28065 F DEBUG   :   NOTE: found under the lib/ directory are readable.
09-25 08:51:44.538 28065 28065 F DEBUG   :   NOTE: On this device, run setenforce 0 to make the libraries readable.
09-25 08:51:44.538 28065 28065 F DEBUG   :   NOTE: Unreadable libraries:
09-25 08:51:44.538 28065 28065 F DEBUG   :   NOTE:   /tmp/user_0/llm/llm_inference_engine_cpu_main
09-25 08:51:44.538 28065 28065 F DEBUG   :       #00 pc 0000000000154bc0  /tmp/user_0/llm/llm_inference_engine_cpu_main
09-25 08:51:44.538 28065 28065 F DEBUG   :       #01 pc 0000000000151518  /tmp/user_0/llm/llm_inference_engine_cpu_main
09-25 08:51:44.538 28065 28065 F DEBUG   :       #02 pc 0000000000070374  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+180) (BuildId: 3cad38c74be79004f5b2927be886e077)
09-25 08:51:44.538 28065 28065 F DEBUG   :       #03 pc 00000000000620a0  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64) (BuildId: 3cad38c74be79004f5b2927be886e077)
```

**Comment by HarishPermal** [comment]
Hi @MuradAlmakhzangi,

Thanks for sharing, can you also share the memory usage details by following the below commands

```bash
adb shell

free -h
```
as shown [here](https://private-user-images.githubusercontent.com/227581675/493722357-4720b98e-2993-4f45-9b5f-f18e3fccc4c8.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg4NzExMzgsIm5iZiI6MTc1ODg3MDgzOCwicGF0aCI6Ii8yMjc1ODE2NzUvNDkzNzIyMzU3LTQ3MjBiOThlLTI5OTMtNGY0NS05YjVmLWYxOGUzZmNjYzRjOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkyNlQwNzEzNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZGUyMmZlZTQ2ZTZjYzBhMWQzYTYwN2Y3YjI4ZDNjODliNjQyYzA1Y2E5NDhkYWIzNTA3ZmJjZTY3ZmM5ZDYxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Jr-qUzgx027qZ7KGDRuirF5ghmlhj2ZjEQZQtAh-Zrs)

**Comment by sicong-li-arm** [comment]
Hi @HarishPermal  thanks for the reply!
We found out that after renaming the signatures and input/output tensors for the gemma3 model, the segfault went away:

```
    -auto* prefill_runner = llm->interpreter->GetSignatureRunner("prefill");
    +auto* prefill_runner = llm->interpreter->GetSignatureRunner("prefill_512");

     -TfLiteTensor* decode_input = decode_runner->input_tensor("args_0");
     -TfLiteTensor* decode_input_pos = decode_runner->input_tensor("args_1");
     +TfLiteTensor* decode_input = decode_runner->input_tensor("tokens");
     +TfLiteTensor* decode_input_pos = decode_runner->input_tensor("input_pos");

    -const TfLiteTensor* logits = decode_runner->output_tensor("output_0");
    +const TfLiteTensor* logits = decode_runner->output_tensor("logits");

    -TfLiteTensor* prefill_input = prefill_runner->input_tensor("args_0");
    -TfLiteTensor* prefill_input_pos = prefill_runner->input_tensor("args_1");
    +TfLiteTensor* prefill_input = prefill_runner->input_tensor("tokens");
    +TfLiteTensor* prefill_input_pos = prefill_runner->input_tensor("input_pos");
```

However we faced another issue which is that if we use the .task format, there's no basic llm utilities support like kv cache, samplers, or attention masking. But if we use the .tflite format, there's an invalid format error as reported here: https://github.com/google-ai-edge/mediapipe/issues/6093

**Comment by HarishPermal** [comment]
Thanks @sicong-li-arm 😊,

I'll try to apply those changes in my environment and let me see how it goes.

**Comment by HarishPermal** [comment]
Hi @MuradAlmakhzangi,

Can you give [this](https://github.com/google-ai-edge/mediapipe/issues/6093#issuecomment-3594673483) a try, and let me know how it went for you.
